tiledb/sm/subarray/relevant_fragments.h

```diff
@@ -0,0 +1,166 @@
+#define TILEDB_RELEVANT_FRAGMENTS_H
+
+#include "tiledb/common/common.h"
+
+#include <vector>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+/**
+ * This class contains the list of relevant fragments. They have either been
+ * already computed or they will return all fragments for the array if not.
+ */
+class RelevantFragments {
+ public:
+  /**
+   * Simple iterator using an index.
+   */
+  struct Iterator {
+    /** Constructs an interator starting at index i. */
+    Iterator(const RelevantFragments& parent, size_t i)
+        : parent_(parent)
+        , i_(i) {
+    }
+
+    /** Returns the object at the current postion. */
+    unsigned operator*() const {
+      return parent_[i_];
+    }
+
+    /** Pre-increment operator. */
+    Iterator& operator++() {
+      i_++;
+      return *this;
+    }
+
+    /** Post increment operator. */
+    Iterator operator++(int) {
+      Iterator tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+
+    /** Equality operator. */
+    friend bool operator==(const Iterator& a, const Iterator& b) {
+      return a.i_ == b.i_;
+    };
+
+    /** Non-equality operator. */
+    friend bool operator!=(const Iterator& a, const Iterator& b) {
+      return a.i_ != b.i_;
+    };
+
+   private:
+    /** Reference to the parent. */
+    const RelevantFragments& parent_;
+
+    /** Current element index. */
+    size_t i_;
+  };
+
+  /**
+   * Constructs a relevant fragment object for an array with `frag_num`
+   * fragments.
+   */
+  RelevantFragments(size_t frag_num)
+      : frag_num_(frag_num)
+      , relevant_fragments_computed_(false) {
+  }
+
+  /** Returns the start iterator for the relevant fragments. */
+  Iterator begin() const {
+    return Iterator(*this, 0);
+  }
+
+  /** Returns the end iterator for the relevant fragments. */
+  Iterator end() const {
+    return Iterator(*this, size());
+  }
+
+  /** Returns the size of the relevant fragments. */
+  inline size_t size() const {
+    return relevant_fragments_computed_ ? computed_relevant_fragments_.size() :
+                                          frag_num_;
+  }
+
+  /**
+   * Returns the size of the computed relevant fragments or 0 if not computed.
+   */
+  inline size_t relevant_fragments_size() const {
+    return computed_relevant_fragments_.size();
+  }
+
+  /** Returns the relevant fragment at index `i`. */
+  inline unsigned operator[](size_t i) const {
+    return relevant_fragments_computed_ ? computed_relevant_fragments_[i] :
+                                          static_cast<unsigned>(i);
+  }
+
+  /** Clears the computed relevant fragments. */
+  inline void clear_computed_relevant_fragments() {
+    relevant_fragments_computed_ = false;
+    computed_relevant_fragments_.clear();
+  }
+
+  /** Reserves the computed relevant fragments vector. */
+  inline void reserve_computed_relevant_fragments(size_t n) {
+    relevant_fragments_computed_ = true;
+    computed_relevant_fragments_.reserve(n);
```

Comment:
I didn't look hard enough at the implementation to know if its a big deal, but if I'm not mistaken, if this reservation isn't followed up by exactly `n` calls to `emplace_computed_relevant_fragments_back` we'll end up with the wrong values in `computed_relevant_fragments_`. I wonder if we should store the number reserved, then turn `relevant_fragments_computed` into a function that's just `return computed_relevant_fragments_.size() == relevant_fragments_reserved_;`?

Reply:
This is a vector reserve so if we don't call emplace back for the number reserved, size will still report only the number of added fragments.

Reply:
Sure, I was worried about this sort of bug though:

```c++
std::vector<int32_t> a;
a.reserve(10);
std::cerr << a[5] << std::endl;
```

Granted, it doesn't matter any more since the code has changed, so nothing to worry about.